towhee/functional/mixins/parallel.py

```diff
@@ -100,41 +114,30 @@ def pmap(self, unary_op, num_worker=None, executor=None):
         Examples:
         >>> from towhee.functional import DataCollection
         >>> import threading
-        >>> stage_1_thread_set = set()
-        >>> stage_2_thread_set = set()
+        >>> stage_1_thread_set = {threading.current_thread().ident}
+        >>> stage_2_thread_set = {threading.current_thread().ident}
         >>> result = (
         ...     DataCollection.range(1000).stream()
         ...     .pmap(lambda x: stage_1_thread_set.add(threading.current_thread().ident), 5)
         ...     .pmap(lambda x: stage_2_thread_set.add(threading.current_thread().ident), 4).to_list()
         ... )
         >>> len(stage_1_thread_set)
-        4
+        5
         >>> len(stage_2_thread_set)
-        3
+        4
         """
-        if executor is None:
-            executor = concurrent.futures.ThreadPoolExecutor(num_worker)
-        num_worker = executor._max_workers  # pylint: disable=protected-access
-        queue = Queue(maxsize=num_worker)
-        loop = asyncio.new_event_loop()
-        flag = True
-
-        def make_task(x):
-
-            def task_wrapper():
-                if isinstance(x, Option):
-                    return x.map(unary_op)
-                else:
-                    return unary_op(x)
-
-            return task_wrapper
+        def inner():
+            nonlocal flag
+            while flag or not queue.empty():
+                yield queue.get()
+            # executor.shutdown()
 
         async def worker():
             buff = []
             for x in self:
                 if len(buff) == num_worker:
                     queue.put(await buff.pop(0))
-                buff.append(loop.run_in_executor(executor, make_task(x)))
+                buff.append(loop.run_in_executor(executor, map_task(x, unary_op)))
             while len(buff) > 0:
                 queue.put(await buff.pop(0))
```

Comment:
As you can see, `queue` is limited by `maxsize`, if not set it will fill the memory when the user has large data. So `queue` is like a producer-consumer mode.

Reply:
Yes, that portion makes sense. We should remove 'buff' and just pump data directly into the queue. For speedup, we can later assign a worker cache to allow the worker to start work on later values.

Reply:
The `buff` is to ensure that the loop execution order is correct, and it is also a producer-consumer mode.

We are supposed to consume in the iterator loop, otherwise, the cache will fill up the memory when the amount of data is large.